Added keep_secret parameter feature

stepanstipl opened the pull request:
This is followup on pull request #92 (sorry for long delay with addressing the issues).

Added feature **keep_secret** to redact how are information outputted to the logs.

Typical use case is to use this when managing sensitive values like passwords, which we do not want in the logs. Using this new parameter keep_secret will prevent this.

Accepted values:

`false` - default, normal behavior
`true` - value will be replaced with "[redacted sensitive information]"
`md5` - value will be replaced with it's md5 hash

tphoney commented:
This needs to be rebased, is this any different to https://docs.puppetlabs.com/references/latest/type.html#file-attribute-show_diff ?

Thanks for putting in the effort and time.

stepanstipl commented:
Hi, yep I have rebased it, it's been a while. I would say it's very similar to show_diff (didn't know about that one). Just the logic here is reverse and in addition allows you to set 'md5' and will output md5 hashes instead of actual values.

Would you prefer to rename this to "show_diff" and align the behaviour with the show_diff from file type? I think it would make sense.

tphoney commented:
Apologies for getting back to this so late. renaming this to "show_diff" makes total sense

tphoney commented:
@stepanstipl thanks for your patience

bmjen commented:
@stepanstipl Thanks for the contribution, have you had a chance to take a look at the comments made by @tphoney and @asasfu ?

stepanstipl commented:
@bmjen, thanks for a poke, it was kind of waiting on my laptop to be pushed.

So, I've made following changes:
- renamed to **show_diff**
- realigned logic to be same as show_diff from file type
- with regards to @asasfu I've put in conversion function to handle boolean values entered as strings or symbols
- updated docs, unit and beaker tests

One note here - as I've tried to follow behaviour of show_diff on file type, it'll change default output. By default global `show_diff` option (puppet_conf) is false, and because we respect this, values will be redacted. I assume for typical use case this would affect only log files, as if you run puppet with `--test` it implies `--show_diff`.

asasfu commented on lib/puppet/type/ini_setting.rb:
```diff
@@ -5,6 +7,19 @@
+  def munge_boolean_md5(value)
+    case value
+    when true, :true, 'true', :yes, 'yes'
+      true
```
Change looks pretty good, take a look here though https://github.com/puppetlabs/puppet/blob/master/lib/puppet/type/macauthorization.rb#L20-L29 you'll notice it should return :true or :false not true or false as the insync? seems to have an issue if it's not a symbol(I don't know why, just tested it and found it to be the correct method)

asasfu commented on lib/puppet/type/ini_subsetting.rb:
```diff
@@ -5,6 +7,19 @@
+  def munge_boolean_md5(value)
+    case value
+    when true, :true, 'true', :yes, 'yes'
+      true
```
Same as other comment, return :true or :false

stepanstipl commented:
@asasfu thanks for comments, updated

jonnytpuppet commented:
@stepanstipl It would be great if you could fix the test failures before this can be merged.

Thanks!

stepanstipl commented:
@jonnytpuppet  I have rebased it onto the current master, the tests seems to be passing now

tphoney commented:
Thanks for the work @stepanstipl  :)
